### PR TITLE
fix in source set compile task doFirst action definition used for semantic db configuration

### PR DESCRIPTION
**Pull request by fsanjuan**
Fix for https://github.com/cosmicsilence/gradle-scalafix/issues/58

### CODE REVIEW

**Comment by aishfenton** [comment]
Not sure if you also want to look at this as part of it too. https://github.com/cosmicsilence/gradle-scalafix/issues/58#issuecomment-987490084

Thankfully the plugin option has an easy workaround (i.e. using Gradle's scalaCompilerPlugin mechanism instead of setting the option). But the `sourceroot` part is still puzzling me. I was hoping maybe it could just be removed (since according to the semanticdb docs, all it does is relativize the paths), but without it the generated `semenaticdb` files aren't even appearing in the build directory. So quite confused by that.

**Comment by fsanjuan** [comment]
> Not sure if you also want to look at this as part of it too. [#58 (comment)](https://github.com/cosmicsilence/gradle-scalafix/issues/58#issuecomment-987490084)
> 
> Thankfully the plugin option has an easy workaround (i.e. using Gradle's scalaCompilerPlugin mechanism instead of setting the option). But the `sourceroot` part is still puzzling me. I was hoping maybe it could just be removed (since according to the semanticdb docs, all it does is relativize the paths), but without it the generated `semenaticdb` files aren't even appearing in the build directory. So quite confused by that.

👋 Apologies for the late reply. I just read your comments there :)

Let's keep this PR to upgrade the plugins and fix that conditional `doFirst` action. We will work on that in a new one :D

**Review by marcelocenerine — DISMISSED** [review]

**Comment by marcelocenerine** [comment]
could you take a look a this [comment](https://github.com/cosmicsilence/gradle-scalafix/issues/58#issuecomment-995246857) before merging this PR?

File: src/main/groovy/io/github/cosmicsilence/scalafix/ScalafixPlugin.groovy [hunk 1]
```diff
@@ -81,12 +81,14 @@ class ScalafixPlugin implements Plugin<Project> {
             scalafixTask.compileOptions.set(project.provider({ sourceSet.compilerOptions }))
             scalafixTask.semanticdbConfigured = extension.semanticdb.autoConfigure.get()
 
-            if (extension.semanticdb.autoConfigure.get()) {
-                // configures the semanticdb compiler plugin during the execution phase, but before the
-                // compile task is executed. This prevents dependencies from being resolved too early
-                sourceSet.compileTask.doFirst {
+            sourceSet.compileTask.doFirst {
```
Comment (marcelocenerine) [inline src/main/groovy/io/github/cosmicsilence/scalafix/ScalafixPlugin.groovy:84]: should we consider replacing the `doFirst` action with a "configure task" as suggested [here](https://docs.gradle.org/current/userguide/common_caching_problems.html#custom_actions)?

File: src/main/groovy/io/github/cosmicsilence/scalafix/ConfigureSemanticdbCompiler.groovy [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package io.github.cosmicsilence.scalafix
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Optional
+import org.gradle.api.tasks.TaskAction
+
+class ConfigureSemanticdbCompiler extends DefaultTask {
```
Comment (marcelocenerine) [inline src/main/groovy/io/github/cosmicsilence/scalafix/ConfigureSemanticdbCompiler.groovy:9]: `ConfigureSemanticdbTask`?

File: src/main/groovy/io/github/cosmicsilence/scalafix/ScalafixPlugin.groovy [hunk 3]
```diff
@@ -50,8 +50,16 @@ class ScalafixPlugin implements Plugin<Project> {
         project.sourceSets.each { SourceSet sourceSet ->
             if (ScalaSourceSet.isScalaSourceSet(project, sourceSet) && !extension.ignoreSourceSets.get().contains(sourceSet.name)) {
                 def scalaSourceSet = new ScalaSourceSet(project, sourceSet)
-                configureTaskForSourceSet(project, scalaSourceSet, IN_PLACE, fixTask, extension)
-                configureTaskForSourceSet(project, scalaSourceSet, CHECK, checkTask, extension)
+                ConfigureSemanticdbCompiler configureSemanticdbCompilerTask = project.tasks.create(
+                        "configure${sourceSet.name.capitalize()}SemanticdbCompilerPlugin", ConfigureSemanticdbCompiler) { task ->
```
Comment (marcelocenerine) [inline src/main/groovy/io/github/cosmicsilence/scalafix/ScalafixPlugin.groovy:54]: - `"configureSemanticdb${sourceSet.name.capitalize()}"`?
- let's add a group/description to it

File: src/main/groovy/io/github/cosmicsilence/scalafix/ScalafixPlugin.groovy [hunk 4]
```diff
@@ -81,33 +90,16 @@ class ScalafixPlugin implements Plugin<Project> {
             scalafixTask.compileOptions.set(project.provider({ sourceSet.compilerOptions }))
             scalafixTask.semanticdbConfigured = extension.semanticdb.autoConfigure.get()
 
+            sourceSet.compileTask.dependsOn configureSemanticdbCompilerTask
```
Comment (marcelocenerine) [inline src/main/groovy/io/github/cosmicsilence/scalafix/ScalafixPlugin.groovy:93]: could we have a comment here explaining why this is done inside `register` and not outside?

File: src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginTest.groovy [hunk 5]
```diff
@@ -335,8 +335,10 @@ class ScalafixPluginTest extends Specification {
         ScalafixTask scalafixTestTask = tasks.scalafixTest // forces plugin configuration
         ScalaCompile compileMainTask = tasks.compileScala
         ScalaCompile compileTestTask = tasks.compileTestScala
-        [compileMainTask, compileTestTask].each { task -> // emulates the task execution
-            task.actions.find { it.displayName.contains('doFirst') }.execute(task)
+
+        [tasks.configureMainSemanticdbCompilerPlugin, tasks.configureTestSemanticdbCompilerPlugin].each { task -> // emulates the task execution
+            assert task.onlyIf.specs.get(1).isSatisfiedBy(task)
+            task.run()
```
Comment (marcelocenerine) [inline src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginTest.groovy:341]: I'm thinking if we could simplify some of these tests by introducing a test class for the new task and moving everything concerning the Semanticdb setup into it. Tests in this class would just check if the dependencies between tasks were created.
Also, there are some tests in `ScalafixPluginFunctionalTest` that may need to be updated

**Review by marcelocenerine — CHANGES_REQUESTED** [review]

**Comment by marcelocenerine** [comment]
Closing this PR since there's been no activity lately. I will resume the work in a new PR.

**Pull request closed without merging**